### PR TITLE
feat: refactor i2gw toolings and add powerful make targets

### PR DESCRIPTION
**Pull request by Xunzhuo**
**What type of PR is this?**

/kind toolings

**What this PR does / why we need it**:

- [ ] refactor toolings: make file, hack script to have an extensible/powerful structure.
- [ ] provide detailed `make help`
- [ ] provide `Golang` area make targets to build and test golang, including:
  - [ ] `make build`: support building host OS and Arch binary.
  - [ ] `make build-multiarch`: support building multiarch binaries of i2gw: `linux/amd64`, `linux/arm64`, `darwin/amd64`, `darwin/arm64`.
  - [ ] `make test`: support running unit test of i2gw.
  - [ ] `make format`: check if go module is clean and runs go mod tidy.
  - [ ] `make clean`: clean output files.
- [ ] provide `Image` area make targets (We do not plan to publish i2gw in image, but this area is needed for developers who want to have local build and push i2gw image and host it privately), including:
  - [ ] `make image`: build host OS and Arch image.
  - [ ] `make image-multiarch`: build multi-arch: `linux/arm64` and `linux/amd64` image with docker buildx.
  - [ ] `make push`: build and push host OS and Arch image.
  - [ ] `make push-multiarch`:  build and push multi-arch: `linux/arm64` and `linux/amd64` binary with docker buildx.
- [ ] provide `Lint` area make targets used for linting codes, including:
  - [ ] `make lint`: Run all linter of code sources, including golint, yamllint, and codespell.
  - [ ] `make lint-deps`: Prepare everything necessary to lint

Sub of #9

**Which issue(s) this PR fixes**:

xref: #21 

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by Xunzhuo** [comment]
xref: https://github.com/kubernetes-sigs/ingress2gateway/issues/21

File: tools/github-actions/setup-deps/action.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+name: setup-deps
+description: Install host system dependencies
+
+runs:
+  using: composite
+  steps:
+    - uses: actions/setup-go@v3
+      with:
+        go-version: 1.20.1
```
Comment (sunjayBhatia) [inline tools/github-actions/setup-deps/action.yaml:9]: could bump this to 1.20.2

File: tools/github-actions/setup-deps/action.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+name: setup-deps
```
[the remark this answers is not in the document]
Comment (Xunzhuo) [reply tools/github-actions/setup-deps/action.yaml:1]: Yes. It is used in other github workflows.

**Comment by robscott** [comment]
@Xunzhuo let us know if you want to follow up on these, for now I'm going to close out in favor of more recent PRs.

**Pull request closed without merging**